File: ticket-booking-admin/src/services/eventSchedule.service.ts
import api from './api';
import { EventSchedule, EventScheduleRequest, ScheduleStatus } from '../types';

class EventScheduleService {
  private getBasePath(): string {
    // Check both sessionStorage and localStorage for user data
    const userStr = sessionStorage.getItem('user') || localStorage.getItem('user');
    
    if (!userStr) {
      return '/api/admin';
    }
    
    try {
      const user = JSON.parse(userStr);
      const role = user.role;
      
      if (role === 'ORGANIZER' || role === 'ROLE_ORGANIZER' ||
          role === 'ORGANIZER_EMPLOYEE' || role === 'ROLE_ORGANIZER_EMPLOYEE') {
        return '/api/organizer';
      } else if (role === 'ADMIN' || role === 'ROLE_ADMIN' || role === 'SUPER_ADMIN' || role === 'ROLE_SUPER_ADMIN') {
        return '/api/admin';
      }
      // Default fallback if role doesn't match any known roles
      return '/api/admin';
    } catch (e) {
      return '/api/admin';
    }
  }

  async getSchedulesForEvent(eventId: string): Promise<EventSchedule[]> {
    const basePath = this.getBasePath();
    const response = await api.get<EventSchedule[]>(`${basePath}/events/${eventId}/schedules`);
    return response.data;
  }

  async getBookableSchedulesForEvent(eventId: string): Promise<EventSchedule[]> {
    const basePath = this.getBasePath();
    const response = await api.get<EventSchedule[]>(`${basePath}/events/${eventId}/schedules/bookable`);
    return response.data;
  }

  async getPublicBookableSchedulesForEvent(eventId: string): Promise<EventSchedule[]> {
    const response = await api.get<EventSchedule[]>(`/api/public/events/${eventId}/schedules/bookable`);
    return response.data;
  }

  async getScheduleById(eventId: string, scheduleId: string): Promise<EventSchedule> {
    const basePath = this.getBasePath();
    const response = await api.get<EventSchedule>(`${basePath}/events/${eventId}/schedules/${scheduleId}`);
    return response.data;
  }

  async createSchedule(eventId: string, request: EventScheduleRequest): Promise<EventSchedule> {
    const basePath = this.getBasePath();
    const response = await api.post<EventSchedule>(`${basePath}/events/${eventId}/schedules`, request);
    return response.data;
  }

  async updateSchedule(eventId: string, scheduleId: string, request: EventScheduleRequest): Promise<EventSchedule> {
    const basePath = this.getBasePath();
    const response = await api.put<EventSchedule>(`${basePath}/events/${eventId}/schedules/${scheduleId}`, request);
    return response.data;
  }

  async deleteSchedule(eventId: string, scheduleId: string): Promise<void> {
    const basePath = this.getBasePath();
    await api.delete(`${basePath}/events/${eventId}/schedules/${scheduleId}`);
  }

  async restoreSchedule(eventId: string, scheduleId: string): Promise<void> {
    const basePath = this.getBasePath();
    await api.post(`${basePath}/events/${eventId}/schedules/${scheduleId}/restore`);
  }

  async permanentlyDeleteSchedule(eventId: string, scheduleId: string): Promise<void> {
    const basePath = this.getBasePath();
    await api.delete(`${basePath}/events/${eventId}/schedules/${scheduleId}/permanent`);
  }

  async changeScheduleStatus(eventId: string, scheduleId: string, status: ScheduleStatus): Promise<EventSchedule> {
    const basePath = this.getBasePath();
    const response = await api.patch<EventSchedule>(
      `${basePath}/events/${eventId}/schedules/${scheduleId}/status?status=${status}`
    );
    return response.data;
  }

  async getSchedulesByDateRange(eventId: string, startDate: string, endDate: string): Promise<EventSchedule[]> {
    const basePath = this.getBasePath();
    const response = await api.get<EventSchedule[]>(
      `${basePath}/events/${eventId}/schedules/range?startDate=${startDate}&endDate=${endDate}`
    );
    return response.data;
  }

  async getTotalCapacity(eventId: string): Promise<number> {
    const basePath = this.getBasePath();
    const response = await api.get<number>(`${basePath}/events/${eventId}/schedules/total-capacity`);
    return response.data;
  }

  async getTotalAvailableSeats(eventId: string): Promise<number> {
    const basePath = this.getBasePath();
    const response = await api.get<number>(`${basePath}/events/${eventId}/schedules/total-available`);
    return response.data;
  }
}

const eventScheduleService = new EventScheduleService();
export default eventScheduleService;

File: ticket-booking-system/src/main/java/com/ticket/ticket_booking_system/service/EventScheduleService.java
package com.ticket.ticket_booking_system.service;

import java.time.LocalDate;
import java.util.List;
import java.util.UUID;

import com.ticket.ticket_booking_system.dto.request.EventScheduleRequest;
import com.ticket.ticket_booking_system.dto.response.EventScheduleResponse;
import com.ticket.ticket_booking_system.entity.EventSchedule.ScheduleStatus;

public interface EventScheduleService {

    // Create new schedule for an event
    EventScheduleResponse createSchedule(UUID eventId, EventScheduleRequest request);

    // Update existing schedule
    EventScheduleResponse updateSchedule(UUID scheduleId, EventScheduleRequest request);

    // Delete schedule (soft delete - move to recycle bin)
    void deleteSchedule(UUID scheduleId);

    // Restore schedule from recycle bin
    void restoreSchedule(UUID scheduleId);

    // Permanently delete schedule
    void permanentlyDeleteSchedule(UUID scheduleId);

    // Get schedule by ID
    EventScheduleResponse getScheduleById(UUID scheduleId);

    // Get all schedules for an event
    List<EventScheduleResponse> getSchedulesForEvent(UUID eventId);

    // Get upcoming/bookable schedules for an event (for customers)
    List<EventScheduleResponse> getBookableSchedulesForEvent(UUID eventId);

    // Get schedules by date range
    List<EventScheduleResponse> getSchedulesByDateRange(UUID eventId, LocalDate startDate, LocalDate endDate);

    // Change schedule status
    EventScheduleResponse changeScheduleStatus(UUID scheduleId, ScheduleStatus status);

    // Reserve seats for a booking
    boolean reserveSeats(UUID scheduleId, int numberOfSeats);

    // Release seats (for cancellations)
    void releaseSeats(UUID scheduleId, int numberOfSeats);

    // Get total capacity across all schedules for an event
    Integer getTotalCapacityForEvent(UUID eventId);

    // Get total available seats across all schedules for an event
    Integer getTotalAvailableSeatsForEvent(UUID eventId);
}
